go-controller/pkg/allocator/pod/pod_annotation.go

```diff
@@ -274,60 +274,65 @@ func allocatePodAnnotationWithRollback(
 	hasIPRequest := network != nil && len(network.IPRequest) > 0
 	hasStaticIPRequest := hasIPRequest && !reallocateIP
 
-	if hasIPAM && hasStaticIPRequest {
-		// for now we can't tell apart already allocated IPs from IPs excluded
-		// from allocation so we can't really honor static IP requests when
-		// there is IPAM as we don't really know if the requested IP should not
-		// be allocated or was already allocated by the same pod
-		err = fmt.Errorf("cannot allocate a static IP request with IPAM for pod %s", podDesc)
-		return
-	}
-
 	// we need to update the annotation if it is missing IPs or MAC
 	needsIPOrMAC := len(tentative.IPs) == 0 && (hasIPAM || hasIPRequest)
 	needsIPOrMAC = needsIPOrMAC || len(tentative.MAC) == 0
 	reallocateOnNonStaticIPRequest := len(tentative.IPs) == 0 && hasIPRequest && !hasStaticIPRequest
 
+	staticTentativePresents := false
 	if len(tentative.IPs) == 0 {
 		if hasIPRequest {
 			tentative.IPs, err = util.ParseIPNets(network.IPRequest)
 			if err != nil {
 				return
 			}
 		}
+	} else if hasStaticIPRequest {
+		staticTentativePresents = true
 	}
 
 	if hasIPAM {
-		if len(tentative.IPs) > 0 {
-			if err = ipAllocator.AllocateIPs(tentative.IPs); err != nil && !ip.IsErrAllocated(err) {
-				err = fmt.Errorf("failed to ensure requested or annotated IPs %v for %s: %w",
-					util.StringSlice(tentative.IPs), podDesc, err)
-				if !reallocateOnNonStaticIPRequest {
-					return
+		if staticTentativePresents {
+			releaseIPs = util.CopyIPNets(tentative.IPs)
+		} else {
+			if len(tentative.IPs) > 0 {
+				err = ipAllocator.AllocateIPs(tentative.IPs)
+				if err != nil {
+					if hasStaticIPRequest {
+						return
+					}
```

Comment:
isn't this redundant ? 

I mean ... 
- staticTentativePresents = hasStaticIPRequest
- we're on a block where staticTentativePresents == false, thus hasStaticIPRequest == false.

How can this be true ?

Reply:
staticTentativePresents is true only if the tentative array length was 0, before assigning to it
so here we are on the very first allocate, which should not be fail no matter what,
it cant have 2nd change as the non static - ipam has

Reply:
maybe it can be refactored later, it is just a quick and dirty poc, didnt even yet thought about / tested everything